### DISCUSSION TITLE
tokio_util: Convert MyStream<T> into MyAsync<Bytes>

### DISCUSSION THREAD

**Discussion by ph633x** [tokio-rs/tokio#3703] [Q&A] [answered]
Hello dear tokio friends!

First of all, thanks for this amazing project.

My scenario might seem a little weird, for which I apologize and am open for suggestions:

--

I have `Decoder` for some type `T`> However, in my decoder, I don't consume the stream. Instead, I just look at the first `Bytes`, parse them into a type `T`, and keep the bytes there "untouched" by not 'splitting` them.

In the usage side, I get a `TcpStream`, frame it with `FramedRead` and my `Decoder`, make a decision based on the value of type `T`, and then convert the `FramedRead` back into a `TcpStream` with `into_inner`. My return type **must** be *something* that implements `AsyncRead` and `AsyncWrite`,  not *necessarily* the actual input `TcpStream`.

The problem is that, `into_inner` discards the bytes I had already "read", which are available via `get_read_buf`.

--

I thought about creating a new type wrapping the `TcpStream` and the `Bytes` in `get_read_buf` and implement `AsyncRead` for it. Upon reading, first takes the leftover bytes, and then read from the inner `TcpStream`.

But I am not sure if that's the best way nor how to best implement such a solution.

--

Would you happen to know the best way to solve this? 

Thank you very much.
Leo.

**Comment by Darksonn** [comment, marked as answer]
You could put the `Bytes` object into a [`Cursor`][2] and combine it with the tcp stream using [`AsyncReadExt::chain`][1].

[1]: https://docs.rs/tokio/1/tokio/io/trait.AsyncReadExt.html#method.chain
[2]: https://doc.rust-lang.org/stable/std/io/struct.Cursor.html

**Reply by ph633x** [reply]
True! That's great idea. Thanks for pointing that out.

On top that:

Would this non-consuming decoder be idiomatic usage of tokio_util?

**Reply by Darksonn** [reply]
If you're only going to read a single element in this way, it would make more sense to do it yourself with the ordinary `read` methods.

**Reply by ph633x** [reply]
Pretty much. Actually, I need to read (or peek) the first n elements. 

But indeed, maybe read_exact and then chain with TCP stream via Cursor.

I like it :).

Thanks.
